### PR TITLE
FOGL-7446 add data source to average rule

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

**Review by pintomax — DISMISSED** [review]

File: include/average.h [hunk 1]
```diff
@@ -32,6 +32,9 @@ class AverageRule: public BuiltinRule
 		void	unlockConfig() { m_configMutex.unlock(); };
 		bool	evaluate(const std::string& asset, const std::string& datapoint, long value);
 		bool	evaluate(const std::string& asset, const std::string& datapoint, double value);
+		std::string
+			getSource() { return m_source; };
```
Comment (ashish-jabble) [inline include/average.h:36]: Better to indent

File: plugin.cpp [hunk 2]
```diff
@@ -165,7 +173,13 @@ string plugin_triggers(PLUGIN_HANDLE handle)
 		  it != triggers.end();
 		  ++it)
 	{
-		ret += "{ \"asset\"  : \"" + (*it).first + "\"";
+		string source = rule->getSource();
+		if (source.compare("Reading") == 0)
```
Comment (ashish-jabble) [inline plugin.cpp]: `Readings`

File: plugin.cpp [hunk 3]
```diff
@@ -165,7 +173,13 @@ string plugin_triggers(PLUGIN_HANDLE handle)
...
+			ret += "{ \"asset\"  : \"" + (*it).first + "\"";
+		if (source.compare("Statistics") == 0)
+			ret += "{ \"statistic\"  : \"" + (*it).first + "\"";
+		if (source.compare("Statistics History") == 0)
+			ret += "{ \"statisticRate\"  : \"" + (*it).first + "\"";
```
Comment (ashish-jabble) [inline plugin.cpp]: Better to add `else if` and then in `else` block - Please catch error (Unknown data source ....)

**Review by ashish-jabble — COMMENTED** [review]
Data Source option is available for Threshold and Average Plugin.
Are we planning to add source in others rule like delta, watchdog etc?

**Comment by MarkRiddoch** [comment]
> Data Source option is available for Threshold and Average Plugin.
> Are we planning to add source in others rule like delta, watchdog etc?

Yes, there is a Jira for this, FOGL-7446, this is the first plugin under the Jira

File: plugin.cpp (hunk 2 above)
Comment (pintomax) [reply plugin.cpp]: "Readings" is the right value

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
@MarkRiddoch https://github.com/fledge-iot/fledge-rule-average/pull/28#discussion_r1109632471 needs to be fixed.

**Comment by MarkRiddoch** [comment]
> 

Yes, it will be introduced across most plugins, it will just take a little time to do.

**Review by pintomax — DISMISSED** [review]

**Comment by ashish-jabble** [comment]
> Yes, it will be introduced across most plugins, it will just take a little time to do.

I am not sure what do you mean here. But I think `Readings` is the right value to compare here https://github.com/fledge-iot/fledge-rule-average/blob/9273ef31dd7949074ad23f361dd11b871dc7657e/plugin.cpp#L177

And Inbuilt Threshold has this check already
https://github.com/fledge-iot/fledge-service-notification/blob/b5a0d0235dbc1b1f1f20d87d1b07e6e77641cca7/C/services/notification/threshold_rule.cpp#L193

**Review by ashish-jabble — APPROVED** [review]

**Pull request merged**